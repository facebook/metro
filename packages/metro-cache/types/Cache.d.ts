/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @oncall react_native
 */

import {CacheStore} from './types';

/**
 * Main cache class. Receives an array of cache instances, and sequentially
 * traverses them to return a previously stored value. It also ensures setting
 * the value in all instances.
 *
 * All get/set operations are logged via Metro's logger.
 */
export default class Cache<T> {
  constructor(stores: ReadonlyArray<CacheStore<T>>);
  get(key: Buffer): Promise<T | null>;
  set(key: Buffer, value: T): void;
  get isDisabled(): boolean;
}
