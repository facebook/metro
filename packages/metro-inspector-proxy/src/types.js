/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 * @oncall react_native
 */

'use strict';

// Page information received from the device. New page is created for
// each new instance of VM and can appear when user reloads React Native
// application.
export type Page = {
  id: string,
  title: string,
  vm: string,
  app: string,
  ...
};

// Chrome Debugger Protocol message/event passed between device and debugger.
export type WrappedEvent = {
  event: 'wrappedEvent',
  payload: {
    pageId: string,
    wrappedEvent: string,
    ...
  },
  ...
};

// Request sent from Inspector Proxy to Device when new debugger is connected
// to particular page.
export type ConnectRequest = {
  event: 'connect',
  payload: {pageId: string, ...},
  ...
};

// Request sent from Inspector Proxy to Device to notify that debugger is
// disconnected.
export type DisconnectRequest = {
  event: 'disconnect',
  payload: {pageId: string, ...},
  ...
};

// Request sent from Inspector Proxy to Device to get a list of pages.
export type GetPagesRequest = {event: 'getPages', ...};

// Response to GetPagesRequest containing a list of page infos.
export type GetPagesResponse = {
  event: 'getPages',
  payload: Array<Page>,
  ...
};

// Union type for all possible messages sent from device to Inspector Proxy.
export type MessageFromDevice =
  | GetPagesResponse
  | WrappedEvent
  | DisconnectRequest;

// Union type for all possible messages sent from Inspector Proxy to device.
export type MessageToDevice =
  | GetPagesRequest
  | WrappedEvent
  | ConnectRequest
  | DisconnectRequest;

// Page description object that is sent in response to /json HTTP request from debugger.
export type PageDescription = {
  id: string,
  description: string,
  title: string,
  faviconUrl: string,
  devtoolsFrontendUrl: string,
  type: string,
  webSocketDebuggerUrl: string,
  ...
};
export type JsonPagesListResponse = Array<PageDescription>;

// Response to /json/version HTTP request from the debugger specifying browser type and
// Chrome protocol version.
export type JsonVersionResponse = {
  Browser: string,
  'Protocol-Version': string,
  ...
};

/**
 * Types were exported from https://github.com/ChromeDevTools/devtools-protocol/blob/master/types/protocol.d.ts
 */

export type SetBreakpointByUrlRequest = {
  id: number,
  method: 'Debugger.setBreakpointByUrl',
  params: {
    lineNumber: number,
    url?: string,
    urlRegex?: string,
    scriptHash?: string,
    columnNumber?: number,
    condition?: string,
  },
};

export type GetScriptSourceRequest = {
  id: number,
  method: 'Debugger.getScriptSource',
  params: {
    scriptId: string,
  },
};

export type GetScriptSourceResponse = {
  scriptSource: string,
  /**
   * Wasm bytecode.
   */
  bytecode?: string,
};

export type ErrorResponse = {
  error: {
    message: string,
  },
};

export type DebuggerRequest =
  | SetBreakpointByUrlRequest
  | GetScriptSourceRequest;
