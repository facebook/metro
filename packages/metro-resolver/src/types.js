/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 * @oncall react_native
 */

'use strict';

export type Result<+TResolution, +TCandidates> =
  | {+type: 'resolved', +resolution: TResolution}
  | {+type: 'failed', +candidates: TCandidates};

export type Resolution = FileResolution | {+type: 'empty'};

export type AssetFileResolution = $ReadOnlyArray<string>;
export type FileResolution =
  | {+type: 'sourceFile', +filePath: string}
  | {+type: 'assetFiles', +filePaths: AssetFileResolution};

export type FileAndDirCandidates = {
  +dir: FileCandidates,
  +file: FileCandidates,
};

/**
 * This is a way to describe what files we tried to look for when resolving
 * a module name as file. This is mainly used for error reporting, so that
 * we can explain why we cannot resolve a module.
 */
export type FileCandidates =
  // We only tried to resolve a specific asset.
  | {+type: 'asset', +name: string}
  // We attempted to resolve a name as being a source file (ex. JavaScript,
  // JSON...), in which case there can be several extensions we tried, for
  // example `/js/foo.ios.js`, `/js/foo.js`, etc. for a single prefix '/js/foo'.
  | {
      +type: 'sourceFile',
      filePathPrefix: string,
      +candidateExts: $ReadOnlyArray<string>,
    };

/**
 * Check existence of a single file.
 */
export type DoesFileExist = (filePath: string) => boolean;
export type IsAssetFile = (fileName: string) => boolean;

/**
 * Given a directory path and the base asset name, return a list of all the
 * asset file names that match the given base name in that directory. Return
 * null if there's no such named asset. `platform` is used to identify
 * platform-specific assets, ex. `foo.ios.js` instead of a generic `foo.js`.
 */
export type ResolveAsset = (
  dirPath: string,
  assetName: string,
  extension: string,
) => ?$ReadOnlyArray<string>;

export type FileContext = $ReadOnly<{
  +doesFileExist: DoesFileExist,
  +isAssetFile: IsAssetFile,
  +nodeModulesPaths: $ReadOnlyArray<string>,
  +preferNativePlatform: boolean,
  +redirectModulePath: (modulePath: string) => string | false,
  +resolveAsset: ResolveAsset,
  +sourceExts: $ReadOnlyArray<string>,
  ...
}>;

export type FileOrDirContext = $ReadOnly<{
  ...FileContext,
  /**
   * This should return the path of the "main" module of the specified
   * `package.json` file, after post-processing: for example, applying the
   * 'browser' field if necessary.
   *
   * FIXME: move the post-processing here. Right now it is
   * located in `node-haste/Package.js`, and fully duplicated in
   * `ModuleGraph/node-haste/Package.js` (!)
   */
  +getPackageMainPath: (packageJsonPath: string) => string,
  ...
}>;

export type HasteContext = $ReadOnly<{
  ...FileOrDirContext,
  /**
   * Given a name, this should return the full path to the file that provides
   * a Haste module of that name. Ex. for `Foo` it may return `/smth/Foo.js`.
   */
  +resolveHasteModule: (name: string) => ?string,
  /**
   * Given a name, this should return the full path to the package manifest that
   * provides a Haste package of that name. Ex. for `Foo` it may return
   * `/smth/Foo/package.json`.
   */
  +resolveHastePackage: (name: string) => ?string,
  ...
}>;

export type ModulePathContext = $ReadOnly<{
  ...FileOrDirContext,
  /**
   * Full path of the module that is requiring or importing the module to be
   * resolved.
   */
  +originModulePath: string,
  ...
}>;

export type ResolutionContext = $ReadOnly<{
  ...HasteContext,
  allowHaste: boolean,
  disableHierarchicalLookup: boolean,
  extraNodeModules: ?{[string]: string, ...},
  originModulePath: string,
  resolveRequest?: ?CustomResolver,
  customResolverOptions: CustomResolverOptions,
  ...
}>;

export type CustomResolutionContext = $ReadOnly<{
  ...ResolutionContext,
  resolveRequest: CustomResolver,
  ...
}>;

export type CustomResolver = (
  context: CustomResolutionContext,
  moduleName: string,
  platform: string | null,
) => Resolution;

export type CustomResolverOptions = {
  __proto__: null,
  +[string]: mixed,
  ...
};
