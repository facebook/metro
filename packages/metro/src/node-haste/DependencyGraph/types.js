/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

'use strict';

import type EventEmitter from 'events';

// TODO(cpojer): Create a jest-types repo.
export type HasteFS = {
  exists(filePath: string): boolean,
  getAllFiles(): Array<string>,
  getDependencies(filePath: string): Array<string>,
  getFileIterator(): Iterator<string>,
  getModuleName(filePath: string): ?string,
  getSha1(string): ?string,
  matchFiles(pattern: RegExp | string): Array<string>,
  ...
};

export type HasteConfig = $ReadOnly<{
  cacheDirectory?: ?string,
  computeDependencies?: ?boolean,
  computeSha1?: ?boolean,
  dependencyExtractor?: ?string | null,
  enableSymlinks?: ?boolean,
  extensions: $ReadOnlyArray<string>,
  forceNodeFilesystemAPI?: ?boolean,
  hasteImplModulePath?: ?string,
  hasteMapModulePath?: ?string,
  ignorePattern?: ?RegExp | ((str: string) => boolean),
  maxWorkers: number,
  mocksPattern?: ?string,
  name: string,
  platforms: $ReadOnlyArray<string>,
  resetCache?: ?boolean,
  retainAllFiles: boolean,
  rootDir: string,
  roots: $ReadOnlyArray<string>,
  skipPackageJson?: ?boolean,
  throwOnModuleCollision?: ?boolean,
  useWatchman?: ?boolean,
  watch?: ?boolean,
  ...
}>;

type ModuleMapItem = {[platform: string]: ModuleMetaData};
type ModuleMetaData = [string, number];
type MockData = Map<string, string>;
type ModuleMapData = Map<string, ModuleMapItem>;
type DuplicatesSet = Map<string, number>;
type DuplicatesIndex = Map<string, Map<string, DuplicatesSet>>;

type RawModuleMap = {
  rootDir: string,
  duplicates: DuplicatesIndex,
  map: ModuleMapData,
  mocks: MockData,
};

// `jest-haste-map`'s interface for ModuleMap.
export type ModuleMap = {
  getModule(
    name: string,
    platform: string | null,
    supportsNativePlatform: ?boolean,
  ): ?string,
  getPackage(
    name: string,
    platform: string | null,
    supportsNativePlatform: ?boolean,
  ): ?string,
  getRawModuleMap(): RawModuleMap,
  ...
};

// Interface as used by node-haste / internal tools, satisfied by JestHasteMap
export interface HasteMap extends EventEmitter {
  build: () => Promise<{hasteFS: HasteFS, moduleMap: ModuleMap}>;
  end: () => void;
  getCacheFilePath: () => string;
}
