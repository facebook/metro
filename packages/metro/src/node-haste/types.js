/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

'use strict';

// TODO(cpojer): Create a jest-types repo.
export type HasteFS = {
  exists(filePath: string): boolean,
  getAllFiles(): Array<string>,
  getFileIterator(): Iterator<string>,
  getModuleName(filePath: string): ?string,
  getSha1(string): ?string,
  matchFiles(pattern: RegExp | string): Array<string>,
};
