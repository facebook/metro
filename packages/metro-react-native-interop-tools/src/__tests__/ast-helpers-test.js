/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @flow strict-local
 * @format
 */

import * as t from '@babel/types';
import {
  isTurboModule,
  getTypeAnnotation,
  getFunctionTypeParameter,
  getFunctionTypeAnnotation,
  getObjectTypeAnnotation,
  getObjectTypeProperty,
  getObjectTypeSpreadProperty,
  getGenericTypeAnnotation,
  getTupleTypeAnnotation,
  getUnionTypeAnnotation,
  getIntersectionTypeAnnotation,
  getArrayTypeAnnotation,
  getStringLiteralTypeAnnotation,
  getNumberLiteralTypeAnnotation,
  getInterfaceExtends,
  getTypeParameters,
  getNameFromID,
  getNodeLoc,
  getBoundarySchemaFromAST,
} from '../ast-helpers.js';
import {parse} from 'hermes-parser';

test('isTurboModule returns true, name is "TurboModule" and typeParams is null', () => {
  expect(
    isTurboModule(t.interfaceExtends(t.identifier('TurboModule'))),
  ).toEqual(true);
});

test('isTurboModule returns false, name is not "TurboModule"', () => {
  expect(
    isTurboModule(t.interfaceExtends(t.identifier('OtherModule'))),
  ).toEqual(false);
});

test('isTurboModule returns false, typeParameters it is not empty', () => {
  expect(
    isTurboModule(
      t.interfaceExtends(
        t.identifier('TurboModule'),
        t.typeParameterInstantiation([t.anyTypeAnnotation()]),
      ),
    ),
  ).toEqual(false);
});

test('getTypeAnnotation, testing BooleanTypeAnnotation', () => {
  expect(getTypeAnnotation(t.booleanTypeAnnotation()).type).toBe(
    t.booleanTypeAnnotation().type,
  );
});

test('getTypeAnnotation, testing NumberTypeAnnotation', () => {
  expect(getTypeAnnotation(t.numberTypeAnnotation()).type).toBe(
    t.numberTypeAnnotation().type,
  );
});

test('getTypeAnnotation, testing StringTypeAnnotation', () => {
  expect(getTypeAnnotation(t.stringTypeAnnotation()).type).toBe(
    t.stringTypeAnnotation().type,
  );
});

test('getTypeAnnotation, testing VoidTypeAnnotation', () => {
  expect(getTypeAnnotation(t.voidTypeAnnotation()).type).toBe(
    t.voidTypeAnnotation().type,
  );
});

test('getTypeAnnotation, testing UnknownTypeAnnotation', () => {
  expect(getTypeAnnotation(t.booleanLiteralTypeAnnotation(true)).type).toBe(
    'UnknownTypeAnnotation',
  );
});

test('getTypeAnnotation, testing NullableTypeAnnotation', () => {
  expect(
    getTypeAnnotation(t.nullableTypeAnnotation(t.anyTypeAnnotation())),
  ).toEqual({
    type: 'NullableTypeAnnotation',
    loc: null,
    typeAnnotation: {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  });
});

test('getFunctionTypeAnnotation, function has a function as parameter', () => {
  const callback: BabelNodeFunctionTypeAnnotation = t.functionTypeAnnotation(
    undefined,
    [],
    undefined,
    t.voidTypeAnnotation(),
  );
  const functionNode: BabelNodeFunctionTypeAnnotation =
    t.functionTypeAnnotation(
      undefined,
      [
        t.functionTypeParam(
          t.identifier('screenShoudBeKeptOn'),
          t.anyTypeAnnotation(),
        ),
        t.functionTypeParam(t.identifier('callback'), callback),
      ],
      undefined,
      t.anyTypeAnnotation(),
    );
  expect(getFunctionTypeAnnotation(functionNode)).toEqual({
    type: 'FunctionTypeAnnotation',
    loc: null,
    params: [
      {
        name: 'screenShoudBeKeptOn',
        typeAnnotation: {
          type: 'AnyTypeAnnotation',
          loc: null,
        },
      },
      {
        name: 'callback',
        typeAnnotation: {
          type: 'FunctionTypeAnnotation',
          loc: null,
          params: [],
          returnTypeAnnotation: {
            type: 'VoidTypeAnnotation',
            loc: null,
          },
        },
      },
    ],
    returnTypeAnnotation: {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  });
});

test('getFunctionTypeParameter, testig basic type parameter', () => {
  const param: BabelNodeFunctionTypeParam = t.functionTypeParam(
    t.identifier('testParam'),
    t.anyTypeAnnotation(),
  );
  expect(getFunctionTypeParameter(param)).toEqual({
    name: 'testParam',
    typeAnnotation: {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  });
});

test('getObjectTypeAnnotation, testing an object with a AnyTypeAnnotation property', () => {
  const property: BabelNodeObjectTypeProperty = t.objectTypeProperty(
    t.identifier('setKeepScreenOn'),
    t.anyTypeAnnotation(),
    t.variance('minus'),
  );
  const objectNode: BabelNodeObjectTypeAnnotation = t.objectTypeAnnotation(
    [property],
    [],
    [],
    [],
    false,
  );
  expect(getObjectTypeAnnotation(objectNode)).toEqual({
    type: 'ObjectTypeAnnotation',
    loc: null,
    properties: [
      {
        loc: null,
        name: 'setKeepScreenOn',
        optional: undefined,
        typeAnnotation: {
          type: 'AnyTypeAnnotation',
          loc: null,
        },
      },
    ],
  });
});

test('getObjectTypeProperty, testing AnyTypeAnnotation property', () => {
  const property: BabelNodeObjectTypeProperty = t.objectTypeProperty(
    t.identifier('testProp'),
    t.anyTypeAnnotation(),
    t.variance('plus'),
  );
  expect(getObjectTypeProperty(property)).toEqual({
    loc: null,
    name: 'testProp',
    optional: undefined,
    typeAnnotation: {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  });
});

test('getObjectTypeSpreadProperty returns unknown type', () => {
  const spreadProperty: BabelNodeObjectTypeSpreadProperty =
    t.objectTypeSpreadProperty(t.anyTypeAnnotation());
  expect(getObjectTypeSpreadProperty(spreadProperty)).toEqual({
    loc: null,
    name: '',
    optional: false,
    typeAnnotation: {
      loc: null,
      type: 'UnknownTypeAnnotation',
    },
  });
});

test('getTupleTypeAnnotation, testing basic tuple', () => {
  const typeNode: BabelNodeTupleTypeAnnotation = t.tupleTypeAnnotation([
    t.anyTypeAnnotation(),
    t.anyTypeAnnotation(),
  ]);
  expect(getTupleTypeAnnotation(typeNode)).toEqual({
    type: 'TupleTypeAnnotation',
    loc: null,
    types: [
      {type: 'AnyTypeAnnotation', loc: null},
      {type: 'AnyTypeAnnotation', loc: null},
    ],
  });
});

test('getGenericTypeAnnotation, testing a generic type', () => {
  const typeNode: BabelNodeGenericTypeAnnotation = t.genericTypeAnnotation(
    t.identifier('testGeneric'),
    t.typeParameterInstantiation([
      t.anyTypeAnnotation(),
      t.anyTypeAnnotation(),
    ]),
  );
  expect(getGenericTypeAnnotation(typeNode)).toEqual({
    type: 'GenericTypeAnnotation',
    loc: null,
    name: 'testGeneric',
    typeParameters: [
      {type: 'AnyTypeAnnotation', loc: null},
      {type: 'AnyTypeAnnotation', loc: null},
    ],
  });
});

test('getUnionTypeAnnotation, testing an union type', () => {
  const typeNode: BabelNodeUnionTypeAnnotation = t.unionTypeAnnotation([
    t.anyTypeAnnotation(),
    t.anyTypeAnnotation(),
  ]);
  expect(getUnionTypeAnnotation(typeNode)).toEqual({
    type: 'UnionTypeAnnotation',
    loc: null,
    types: [
      {type: 'AnyTypeAnnotation', loc: null},
      {type: 'AnyTypeAnnotation', loc: null},
    ],
  });
});

test('getIntersectionTypeAnnotation, testing an intersection type', () => {
  const typeNode: BabelNodeIntersectionTypeAnnotation =
    t.intersectionTypeAnnotation([
      t.anyTypeAnnotation(),
      t.anyTypeAnnotation(),
    ]);
  expect(getIntersectionTypeAnnotation(typeNode)).toEqual({
    type: 'IntersectionTypeAnnotation',
    loc: null,
    types: [
      {type: 'AnyTypeAnnotation', loc: null},
      {type: 'AnyTypeAnnotation', loc: null},
    ],
  });
});

test('getArrayTypeAnnotation, testing an array of AnyTypeAnnotation', () => {
  const arrayNode: BabelNodeArrayTypeAnnotation = t.arrayTypeAnnotation(
    t.anyTypeAnnotation(),
  );
  expect(getArrayTypeAnnotation(arrayNode)).toEqual({
    type: 'ArrayTypeAnnotation',
    loc: null,
    elementType: {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  });
});

test('getNumberLiteralTypeAnnotation, testing NumberLiteralType', () => {
  const typeNode: BabelNodeNumberLiteralTypeAnnotation =
    t.numberLiteralTypeAnnotation(4);
  expect(getNumberLiteralTypeAnnotation(typeNode)).toEqual({
    type: 'NumberLiteralTypeAnnotation',
    loc: null,
    value: 4,
  });
});

test('getStringLiteralTypeAnnotation, testing StringLiteralType', () => {
  const typeNode: BabelNodeStringLiteralTypeAnnotation =
    t.stringLiteralTypeAnnotation('test');
  expect(getStringLiteralTypeAnnotation(typeNode)).toEqual({
    type: 'StringLiteralTypeAnnotation',
    loc: null,
    value: 'test',
  });
});

test('getTypeParameters, testing AnyTypeAnnotation parameter', () => {
  const params: Array<BabelNodeFlowType> = [t.anyTypeAnnotation()];
  expect(getTypeParameters(params)).toEqual([
    {
      type: 'AnyTypeAnnotation',
      loc: null,
    },
  ]);
});

test('getInterfaceExtends, testing interface with no parameters', () => {
  const interfaceNode: BabelNodeInterfaceExtends = t.interfaceExtends(
    t.identifier('test'),
    undefined,
  );
  expect(getInterfaceExtends(interfaceNode)).toEqual({
    loc: null,
    name: 'test',
    typeParameters: [],
  });
});

test('getNameFromID, testing BabelNodeIdentifier', () => {
  const node: BabelNodeIdentifier = t.identifier('test');
  expect(getNameFromID(node)).toBe('test');
});

test('getNameFromID, testing BabelNodeQualifiedTypeIdentifier', () => {
  const node: BabelNodeQualifiedTypeIdentifier = t.qualifiedTypeIdentifier(
    t.identifier('test'),
    t.identifier('testQualifier'),
  );
  expect(getNameFromID(node)).toBe('test');
});

test('getNameFromID, testing BabelNodeStringLiteral', () => {
  const node: BabelNodeStringLiteral = t.stringLiteral('test');
  expect(getNameFromID(node)).toBe('test');
});

test('getNodeLoc, testing basic loc', () => {
  const ast: BabelNodeFile = parse('test', {
    babel: true,
    sourceType: 'module',
    sourceFilename: 'test.js',
  });
  expect(getNodeLoc(ast.loc)).toEqual({
    start: {
      line: 1,
      column: 0,
    },
    end: {
      line: 1,
      column: 4,
    },
  });
});

test('getNodeLoc, testing undefined loc', () => {
  expect(getNodeLoc(undefined)).toBe(null);
});

test('getBoundarySchemaFromAST, integration test', () => {
  const code: string = `
    export interface Spec extends TurboModule {
      +getConstants: () => {|
        +testGeneric0: Object,
        +testUnion0?: string | string,
        +testIntersection0?: string & string,
        +testTuple0?: [string, number],
        +scale?: number,
        +isSimulator?: boolean,
        +majorOsVersion?: number,
        +isTablet?: boolean,
        +deviceID: string,
      |};
      +setKeepScreenOn: (screenShouldBeKeptOn: boolean) => void;
    }

    const NativeModule: ?Spec = TurboModuleRegistry.get<Spec>('DeviceManager');

    let NativeDeviceManager: ?Spec = null;
    let constants = null;

    if (NativeModule != null) {
      NativeDeviceManager = {
        getConstants(): {|
          +scale?: number,
          +isSimulator?: boolean,
          +majorOsVersion?: number,
          +isTablet?: boolean,
          +deviceID: string,
        |} {
          if (constants == null) {
            constants = NativeModule.getConstants();
          }
          return constants;
        },
        setKeepScreenOn(screenShouldBeKeptOn: boolean): void {
          NativeModule.setKeepScreenOn(screenShouldBeKeptOn);
        },
      };
    }`;
  const ast: BabelNodeFile = parse(code, {
    babel: true,
    sourceType: 'module',
    sourceFilename: 'NativeDeviceManager.js',
  });
  expect(getBoundarySchemaFromAST(ast, 'test.js')).toMatchSnapshot();
});
