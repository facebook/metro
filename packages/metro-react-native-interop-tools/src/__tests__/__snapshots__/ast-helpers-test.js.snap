// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBoundarySchemaFromAST, integration test 1`] = `
Object {
  "source": "test.js",
  "typegenSchema": Object {
    "Spec": Object {
      "typeAnnotation": Object {
        "extends": Array [
          Object {
            "loc": Object {
              "end": Object {
                "column": 45,
                "line": 2,
              },
              "start": Object {
                "column": 34,
                "line": 2,
              },
            },
            "name": "TurboModule",
            "typeParameters": Array [],
          },
        ],
        "innerType": Object {
          "loc": Object {
            "end": Object {
              "column": 5,
              "line": 15,
            },
            "start": Object {
              "column": 46,
              "line": 2,
            },
          },
          "properties": Array [
            Object {
              "loc": Object {
                "end": Object {
                  "column": 8,
                  "line": 13,
                },
                "start": Object {
                  "column": 6,
                  "line": 3,
                },
              },
              "name": "getConstants",
              "optional": false,
              "typeAnnotation": Object {
                "loc": Object {
                  "end": Object {
                    "column": 8,
                    "line": 13,
                  },
                  "start": Object {
                    "column": 21,
                    "line": 3,
                  },
                },
                "params": Array [],
                "returnTypeAnnotation": Object {
                  "loc": Object {
                    "end": Object {
                      "column": 8,
                      "line": 13,
                    },
                    "start": Object {
                      "column": 27,
                      "line": 3,
                    },
                  },
                  "properties": Array [
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 29,
                          "line": 4,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 4,
                        },
                      },
                      "name": "testGeneric0",
                      "optional": false,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 29,
                            "line": 4,
                          },
                          "start": Object {
                            "column": 23,
                            "line": 4,
                          },
                        },
                        "name": "Object",
                        "type": "GenericTypeAnnotation",
                        "typeParameters": Array [],
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 37,
                          "line": 5,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 5,
                        },
                      },
                      "name": "testUnion0",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 37,
                            "line": 5,
                          },
                          "start": Object {
                            "column": 22,
                            "line": 5,
                          },
                        },
                        "type": "UnionTypeAnnotation",
                        "types": Array [
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 28,
                                "line": 5,
                              },
                              "start": Object {
                                "column": 22,
                                "line": 5,
                              },
                            },
                            "type": "StringTypeAnnotation",
                          },
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 37,
                                "line": 5,
                              },
                              "start": Object {
                                "column": 31,
                                "line": 5,
                              },
                            },
                            "type": "StringTypeAnnotation",
                          },
                        ],
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 44,
                          "line": 6,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 6,
                        },
                      },
                      "name": "testIntersection0",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 44,
                            "line": 6,
                          },
                          "start": Object {
                            "column": 29,
                            "line": 6,
                          },
                        },
                        "type": "IntersectionTypeAnnotation",
                        "types": Array [
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 35,
                                "line": 6,
                              },
                              "start": Object {
                                "column": 29,
                                "line": 6,
                              },
                            },
                            "type": "StringTypeAnnotation",
                          },
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 44,
                                "line": 6,
                              },
                              "start": Object {
                                "column": 38,
                                "line": 6,
                              },
                            },
                            "type": "StringTypeAnnotation",
                          },
                        ],
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 38,
                          "line": 7,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 7,
                        },
                      },
                      "name": "testTuple0",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 38,
                            "line": 7,
                          },
                          "start": Object {
                            "column": 22,
                            "line": 7,
                          },
                        },
                        "type": "TupleTypeAnnotation",
                        "types": Array [
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 29,
                                "line": 7,
                              },
                              "start": Object {
                                "column": 23,
                                "line": 7,
                              },
                            },
                            "type": "StringTypeAnnotation",
                          },
                          Object {
                            "loc": Object {
                              "end": Object {
                                "column": 37,
                                "line": 7,
                              },
                              "start": Object {
                                "column": 31,
                                "line": 7,
                              },
                            },
                            "type": "NumberTypeAnnotation",
                          },
                        ],
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 23,
                          "line": 8,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 8,
                        },
                      },
                      "name": "scale",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 23,
                            "line": 8,
                          },
                          "start": Object {
                            "column": 17,
                            "line": 8,
                          },
                        },
                        "type": "NumberTypeAnnotation",
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 30,
                          "line": 9,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 9,
                        },
                      },
                      "name": "isSimulator",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 30,
                            "line": 9,
                          },
                          "start": Object {
                            "column": 23,
                            "line": 9,
                          },
                        },
                        "type": "BooleanTypeAnnotation",
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 32,
                          "line": 10,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 10,
                        },
                      },
                      "name": "majorOsVersion",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 32,
                            "line": 10,
                          },
                          "start": Object {
                            "column": 26,
                            "line": 10,
                          },
                        },
                        "type": "NumberTypeAnnotation",
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 27,
                          "line": 11,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 11,
                        },
                      },
                      "name": "isTablet",
                      "optional": true,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 27,
                            "line": 11,
                          },
                          "start": Object {
                            "column": 20,
                            "line": 11,
                          },
                        },
                        "type": "BooleanTypeAnnotation",
                      },
                    },
                    Object {
                      "loc": Object {
                        "end": Object {
                          "column": 25,
                          "line": 12,
                        },
                        "start": Object {
                          "column": 8,
                          "line": 12,
                        },
                      },
                      "name": "deviceID",
                      "optional": false,
                      "typeAnnotation": Object {
                        "loc": Object {
                          "end": Object {
                            "column": 25,
                            "line": 12,
                          },
                          "start": Object {
                            "column": 19,
                            "line": 12,
                          },
                        },
                        "type": "StringTypeAnnotation",
                      },
                    },
                  ],
                  "type": "ObjectTypeAnnotation",
                },
                "type": "FunctionTypeAnnotation",
              },
            },
            Object {
              "loc": Object {
                "end": Object {
                  "column": 63,
                  "line": 14,
                },
                "start": Object {
                  "column": 6,
                  "line": 14,
                },
              },
              "name": "setKeepScreenOn",
              "optional": false,
              "typeAnnotation": Object {
                "loc": Object {
                  "end": Object {
                    "column": 63,
                    "line": 14,
                  },
                  "start": Object {
                    "column": 24,
                    "line": 14,
                  },
                },
                "params": Array [
                  Object {
                    "loc": Object {
                      "end": Object {
                        "column": 54,
                        "line": 14,
                      },
                      "start": Object {
                        "column": 25,
                        "line": 14,
                      },
                    },
                    "name": "screenShouldBeKeptOn",
                    "optional": false,
                    "typeAnnotation": Object {
                      "loc": Object {
                        "end": Object {
                          "column": 54,
                          "line": 14,
                        },
                        "start": Object {
                          "column": 47,
                          "line": 14,
                        },
                      },
                      "type": "BooleanTypeAnnotation",
                    },
                  },
                ],
                "returnTypeAnnotation": Object {
                  "loc": Object {
                    "end": Object {
                      "column": 63,
                      "line": 14,
                    },
                    "start": Object {
                      "column": 59,
                      "line": 14,
                    },
                  },
                  "type": "VoidTypeAnnotation",
                },
                "type": "FunctionTypeAnnotation",
              },
            },
          ],
          "type": "ObjectTypeAnnotation",
        },
        "type": "InterfaceDeclarationTypeAnnotation",
      },
    },
  },
}
`;
